packages/langium/src/validation/document-validator.ts

```diff
@@ -69,25 +80,35 @@ export class DefaultDocumentValidator {
         }
 
         // Process custom validations
-        diagnostics.push(...this.validateAst(parseResult.value, document));
+        diagnostics.push(...await this.validateAst(parseResult.value, document, cancelToken));
+
+        if (cancelToken.isCancellationRequested) {
+            return undefined;
+        }
 
         return diagnostics;
     }
 
-    protected validateAst(rootNode: AstNode, document: LangiumDocument): Diagnostic[] {
+    protected async validateAst(rootNode: AstNode, document: LangiumDocument, cancelToken: CancellationToken): Promise<Diagnostic[]> {
         const validationItems: Diagnostic[] = [];
         const acceptor: ValidationAcceptor = <N extends AstNode>(severity: 'error' | 'warning' | 'info' | 'hint', message: string, info: DiagnosticInfo<N>) => {
             validationItems.push(this.toDiagnostic(severity, message, info, document));
         };
 
-        const runChecks = (node: AstNode) => {
+        const runChecks = async (node: AstNode) => {
             const checks = this.validationRegistry.getChecks(node.$type);
             for (const check of checks) {
-                check(node, acceptor);
+                await check(node, acceptor);
             }
         };
         runChecks(rootNode);
-        streamAllContents(rootNode).map(c => c.node).forEach(runChecks);
+        for (const node of streamAllContents(rootNode).map(c => c.node)) {
+            await delayNextTickAfter();
+            if (cancelToken.isCancellationRequested) {
+                break;
+            }
+            await runChecks(node);
```

Comment:
In case a validation returns a promise, we could start more validations in parallel.